rfcs/20240312-standardize-tan-customcall-tuplecollective-ops.md-ops.md

```diff
@@ -0,0 +1,182 @@
+
+#### Inputs
+
+| Label | Name      | Type                                                                    | Constraints |
+|-------|-----------|-------------------------------------------------------------------------|-------------|
+| (I1)  | `operand` | tensor of floating-point or complex type or per-tensor quantized tensor | (C1)        |
+
+#### Outputs
+
+| Name     | Type                                                                    | Constraints |
+|----------|-------------------------------------------------------------------------|-------------|
+| `result` | tensor of floating-point or complex type or per-tensor quantized tensor | (C1)        |
+
+#### Constraints
+
+* (C1) `baseline_type(operand) = baseline_type(result)`.
+
+#### Examples
+
+```mlir
+// %operand: [-1.0, 0.0, 1.0]
+%result = "stablehlo.tan"(%operand) : (tensor<3xf32>) -> tensor<3xf32>
+// %result: [-0.76159416, 0.0, 0.76159416]
+```
+
+### custom_call
+
+#### Semantics
+
+Encapsulates an implementation-defined operation `call_target_name` that takes
+`inputs` and `called_computations` and produces `results`. `has_side_effect`,
+`backend_config` and `api_version` may be used to provide additional
+implementation-defined metadata.
+
+#### Inputs
+
+| Label | Name                  | Type                                          |
+|-------|-----------------------|-----------------------------------------------|
+| (I1)  | `inputs`              | variadic number of values                     |
+| (I2)  | `call_target_name`    | constant of type `string`                     |
+| (I3)  | `has_side_effect`     | constant of type `i1`                         |
+| (I4)  | `backend_config`      | constant of type `string`                     |
+| (I5)  | `api_version`         | constant of type `si32`                       |
+| (I6)  | `called_computations` | variadic number of constants of type `string` |
+
+#### Outputs
+
+| Name      | Type                      |
+|-----------|---------------------------|
+| `results` | variadic number of values |
+
+#### Examples
+
+```mlir
+%results = "stablehlo.custom_call"(%input0) {
+  call_target_name = "foo",
+  has_side_effect = false,
+  backend_config = "bar",
+  api_version = 1 : i32,
+  called_computations = [@foo]
+} : (tensor<f64>) -> tensor<f64>
+```
+
+### all_gather
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, concatenates the values
+of the `operands` tensors from each process along `all_gather_dim` and produces a
+`results` tensors.
+
+The operation splits the StableHLO process grid into `process_groups` which is
+defined as follows:
+
+* `cross_replica(replica_groups)`
+  if `channel_id <= 0 and use_global_device_ids = false`.
+* `cross_replica_and_partition(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = false`.
+* `flattened_ids(replica_groups)`
+  if `channel_id > 0 and use_global_device_ids = true`.
+
+Afterwards, within each `process_group`:
```

Comment:
within each `process_group` and for each `i` in `[0, size(operands))`:

* `operands[i]@receiver = [operand[i]@sender for sender in process_group]` for all
  `receiver` in `process_group`.
* `results[i]@process = concatenate(operands[i]@process, all_gather_dim)` for all
  `process` in `process_group`.